dspace-api/src/test/java/org/dspace/discovery/MetadataFieldIndexFactoryImplTest.java

```diff
@@ -0,0 +1,93 @@
+/**
+ * The contents of this file are subject to the license and copyright
+ * detailed in the LICENSE and NOTICE files at the root of the source
+ * tree and available online at
+ *
+ * http://www.dspace.org/license/
+ */
+package org.dspace.discovery;
+
+import static org.junit.Assert.assertTrue;
+
+import org.apache.solr.common.SolrInputDocument;
+import org.dspace.AbstractUnitTest;
+import org.dspace.content.MetadataField;
+import org.dspace.content.MetadataSchema;
+import org.dspace.content.factory.ContentServiceFactory;
+import org.dspace.content.service.MetadataFieldService;
+import org.dspace.content.service.MetadataSchemaService;
+import org.dspace.discovery.indexobject.IndexableMetadataField;
+import org.dspace.discovery.indexobject.MetadataFieldIndexFactoryImpl;
+import org.junit.Test;
+
+/**
+ * Test class for {@link MetadataFieldIndexFactoryImpl}
+ *
+ * @author Maria Verdonck (Atmire) on 23/07/2020
+ */
+public class MetadataFieldIndexFactoryImplTest extends AbstractUnitTest {
+    private MetadataSchemaService metadataSchemaService =
+        ContentServiceFactory.getInstance().getMetadataSchemaService();
+    private MetadataFieldService metadataFieldService = ContentServiceFactory.getInstance().getMetadataFieldService();
+
+    private String schemaName = "schema1";
+    private String elemName1 = "elem1";
+    private String elemName2 = "elem2";
+    private String qualName1 = "qual1";
+
+    private MetadataSchema schema;
+    private MetadataField field1;
+    private MetadataField field2;
+
+    @Test
+    public void test_buildDocument_withQualifier() throws Exception {
```

Comment:
Would it be possible to add a test (either here or in a separate test file) that proves that *newly added or updated* `MetadataField` objects will be automatically reindexed?

I should know the answer to this, but I'm worried that our `MetadataField` index might not be automatically updated.  Is it being updated whenever a new field is added (or an existing field is updated)?

Reply:
@tdonohue Good call that we did not have a test for this. This indeed was not automatically updated on updates of the metadata fields. The tests have been updated to make sure that the index-discovery isn't run during the test (Which was the case previously, making sure that the fields were indexed).
Updates of the metafataFields trigger the indexing of those fields. (As you might have noticed, this now fails the build because of hibernate problems, which will need to be looked into)
To make sure only metadatafields are updated, the [following method](https://github.com/atmire/DSpace/blob/w2p-71752_Metadata-field-search/dspace-api/src/main/java/org/dspace/discovery/SolrServiceImpl.java#L304) is called 

That being said, we would need to find a **more generic solution** to make sure that, when an indexable object has been changed, that it is also actually indexed automatically.
By default, the consumers catch/handle these events, but that does not take into account that more IndexableObject implementation are created in the future.

Reply:
this is a very nice PR but I worry if we are adding a too risky change compared to the benefit of the feature itself. I don't like the idea to index things in two different way, a sync call here and the event consumer for all the other stuff.
Maybe a solution could be to raise an Event about the `SiteObject` and have the consumer to process such event. This would be similar to how the workflow/tasks are indexed due to an event on the underline `Item`. In the detailed information we could eventually decide to pass the information `METADATAFIELD:ID` to allow a more granular control instead than force to reindex all the metadata each time

Reply:
@jonas-atmire : I agree with @abollini's notes above.  Rather than trying to find a *new way* to automatically update the MetadataField index each time a field is added/removed/updated, let's try to rework this to use the existing consumer-based approach.

So, as @abollini suggested, we could consider this a Site-level event (as Metadata Fields and Schemas arguably "belong" to the entire Site, and not an individual object in the Site).  So, that may mean we need to either create a new `SiteConsumer` which can be notified of MetadataField changes... *or*, perhaps we just update the existing [`IndexEventConsumer`](https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/discovery/IndexEventConsumer.java) to support the `Site` object as well.  If we choose the latter, we may be able to trigger a `Event.MODIFY` or `Event.MODIFY_METADATA` on the `Site` object, passing it the `METADATAFIELD:[ID]` to trigger a reindex of that single field, or perhaps `METADATASCHEMA:[ID]` to reindex all fields in a schema.